Fixed bug of determining availability of functions ‘pthread_getconcur…

Vic-Min opened the pull request:
…rency’, ‘pthread_setconcurrency’, ‘pthread_setaffinity_np’ in ‘pthread’ library

CMake cannot find 'pthread_getconcurrency’, ‘pthread_setconcurrency’ and ‘pthread_setaffinity_np' functions in 'pthread' library because when linking internal cmake test did’t specifed ‘-l pthread’.

openscenegraph commented:
Could you provide context for this problem, it's not something I've seen reported as an issue so would like to know where the problems occurred and why before I can review the validity of the changes.  Thanks.

Vic-Min commented:
I build the latest version from the master branch:
```
$ cmake \
-D CMAKE_INSTALL_PREFIX:PATH=/usr \
-D BUILD_OSG_APPLICATIONS:BOOL=ON \
-D BUILD_OSG_EXAMPLES:BOOL=OFF \
-D CMAKE_BUILD_TYPE:STRING=Debug \
../OpenSceneGraph-3.6

```
In the output I see:
```
-- Looking for pthread_create in pthread - found
…
-- Looking for pthread_setconcurrency
-- Looking for pthread_setconcurrency - not found
-- Looking for pthread_getconcurrency
-- Looking for pthread_getconcurrency - not found
-- Looking for pthread_setaffinity_np
-- Looking for pthread_setaffinity_np - not found
-- Performing Test HAVE_THREE_PARAM_SCHED_SETAFFINITY
-- Performing Test HAVE_THREE_PARAM_SCHED_SETAFFINITY – Success
```

In the CMakeFiles/CMakeError.log file I see:

```
Determining if the function pthread_setaffinity_np exists failed with the following output:
…
/usr/bin/cc   -DCHECK_FUNCTION_EXISTS=pthread_setaffinity_np    CMakeFiles/cmTryCompileExec1651159724.dir/CheckFunctionExists.c.o  -o cmTryCompileExec1651159724 -rdynamic 
CMakeFiles/cmTryCompileExec1651159724.dir/CheckFunctionExists.c.o: In function `main':
CheckFunctionExists.c:(.text+0x15): undefined reference to `pthread_setaffinity_np'
collect2: error: ld returned 1 exit status

```
Changed the src/OpenThreads/pthreads file.
Run:
`$ cmake …`

In the output I see:
```
-- Looking for pthread_setconcurrency
-- Looking for pthread_setconcurrency - found
-- Looking for pthread_getconcurrency
-- Looking for pthread_getconcurrency - found
-- Looking for pthread_setaffinity_np
-- Looking for pthread_setaffinity_np - found
-- Performing Test COMPILES_PTHREAD_SETAFFINITY_NP
-- Performing Test COMPILES_PTHREAD_SETAFFINITY_NP – Success

```
In the CMakeFiles/CMakeOutput.log file I see:

```
Determining if the function pthread_setaffinity_np exists passed with the following output:
…
/usr/bin/cc   -DCHECK_FUNCTION_EXISTS=pthread_setaffinity_np    CMakeFiles/cmTryCompileExec2939878863.dir/CheckFunctionExists.c.o  -o cmTryCompileExec2939878863 -rdynamic -lpthread 

```
Added **-lpthread** (In the code section, I could not highlight it in bold)


My environment

```
$ uname -a
Linux AstraOrel-11 4.2.0-23-generic #28astra39 SMP Thu Apr 7 07:20:56 MSK 2016 x86_64 GNU/Linux

$ gcc --version
gcc (Debian 4.7.2-5) 4.7.2

$ cmake --version
cmake version 2.8.12.2

$ /lib/x86_64-linux-gnu/libc.so.6
GNU C Library (Ubuntu EGLIBC 2.15-0ubuntu10.13astra1) stable release version 2.15, by Roland McGrath et al.
Copyright (C) 2012 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.
Compiled by GNU CC version 4.7.2.
Compiled on a Linux 4.2.6 system on 2016-04-07.
Available extensions:
        crypt add-on version 2.1 by Michael Glad and others
        GNU Libidn by Simon Josefsson
        Native POSIX Threads Library by Ulrich Drepper et al
        BIND-8.2.3-T5B
libc ABIs: UNIQUE IFUNC

$ /lib/x86_64-linux-gnu/libpthread.so.0
Native POSIX Threads Library by Ulrich Drepper et al
Copyright (C) 2006 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.
Forced unwind support included.

```

openscenegraph commented:
Thanks for the info, this was really helpful in giving context and how to reproduce the issue.  I have both reproduced the problem and confirmed that your changes fix the issue.  

I have merged your fix to OSG master and the OpenSceneGraph-3.6 and OpenSceneGraph-3.4 branches.
